src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,426 @@
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            if (!LooksLikeNodeInArgumentList(currentNode, caret, isCaretAtEndOfLine, syntaxFacts))
+            {
+                return;
+            }
+
+            // verify all delimeters exist until you reach statement syntax that requires a semicolon
+            while (!IsStatementOrFieldDeclaration(currentNode, syntaxFacts))
+            {
+                if (!ClosingDelimiterExistsIfNeeded(currentNode))
+                {
+                    // A required delimiter is missing; do not treat semicolon as statement completion
+                    return;
+                }
+
+                if (currentNode.Parent == null)
+                {
+                    return;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            // if the statement syntax itself requires a closing delimeter, verify it is there
+            if (StatementClosingDelimiterIsMissing(currentNode))
+            {
+                // Example: missing final `)` in `do { } while (x$$`
+                return;
+            }
+
+            var semicolonPosition = GetSemicolonLocation(root, currentNode, caretPosition);
+
+            // Place cursor after the statement
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(semicolonPosition));
+        }
+
+        private int GetSemicolonLocation(SyntaxNode root, SyntaxNode currentNode, int caretPosition)
+        {
+            int end;
+            if (currentNode.IsKind(SyntaxKind.ForStatement))
+            {
+                // in for statements, semicolon can go after initializer or after condition, depending on where the caret is located
+                var forStatementSyntax = (ForStatementSyntax)currentNode;
+                if (CaretIsInForStatementCondition(caretPosition, forStatementSyntax))
+                {
+                    end = forStatementSyntax.Condition.FullSpan.End;
+                }
+                else if (CaretIsInForStatementDeclaration(caretPosition, forStatementSyntax))
+                {
+                    end = forStatementSyntax.Declaration.FullSpan.End;
+                }
+                else
+                {
+                    // Should not be reachable because we returned earlier for this case
+                    end = 0;
+                    Debug.Fail("Complete statement not supported on For statement iterator");
+                }
+            }
+            else
+            {
+                end = currentNode.Span.End;
+            }
+
+            // If a node's semicolon is missing, the trailing trivia (including new line) is associated with the last token.
+            // To avoid placing the semicolon on the next line or after comments, we need to adjust the position.
+            return root.FindToken(end).GetPreviousToken().Span.End;
```

Comment:
i don't really understand this.  You said:

>  If a node's semicolon is missing, the trailing trivia (including new line) is associated with the last token.

Isn't that *precisely* what `currentNode.Span.End` already is?  `Span` should not be including trailing-trivia.  So this is just recomputing the same 'end' position you already have.

Reply:
➡️ No. `currentNode.Span.End` only ignores the trailing trivia of the last token in the node. The last token in the node is the inserted semicolon token, which is positioned after the trivia we want to ignore.